subprojects/scala/src/test/groovy/org/gradle/api/internal/tasks/scala/DefaultScalaJavaJointCompileSpecFactoryTest.groovy

```diff
@@ -22,17 +22,29 @@ import org.gradle.api.internal.tasks.compile.ForkingJavaCompileSpec
 import org.gradle.api.model.ObjectFactory
 import org.gradle.api.tasks.compile.CompileOptions
 import org.gradle.internal.jvm.Jvm
+import org.gradle.internal.os.OperatingSystem
 import org.gradle.jvm.toolchain.JavaInstallationMetadata
 import org.gradle.jvm.toolchain.JavaLanguageVersion
+import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
+import org.junit.Rule
 import spock.lang.Specification
 import spock.lang.Subject
 
 @Subject(DefaultScalaJavaJointCompileSpecFactory)
 class DefaultScalaJavaJointCompileSpecFactoryTest extends Specification {
 
+    @Rule
+    TestNameTestDirectoryProvider tmpDir = new TestNameTestDirectoryProvider(getClass())
```

Comment:
If you create tests with that rule, you should annotate them with `@CleanupTestDirectory` I think.

Reply:
It seems `@CleanupTestDirectory` is needed to *keep* the test directory in case of test failures, which is slightly counter-intuitive 🙂 

We use it for all integration tests where it's important to be able to look at the generated build scripts or outputs of tasks, but less so for the unit tests. In this test in particular we use it only to mock the existance of another JDK, so it's okay if it always gets cleaned up.